go/pkg/cmpfeeds/compare_ws_grpc_block_feeds.go

```diff
@@ -0,0 +1,614 @@
+	}
+
+	if blockSeenByBothFeedsGrpcGatewayFirst != 0 {
+		blockReceivedByGrpcGatewayFirstAvgDelta = blockReceivedByGrpcGatewayFirstTotalDelta / blockSeenByBothFeedsGrpcGatewayFirst
+	}
+
+	if newBlockSeenByBothFeeds != 0 {
+		blockPercentageSeenByGatewayFirst = float64(blockSeenByBothFeedsGrpcGatewayFirst) / float64(newBlockSeenByBothFeeds)
+	}
+
+	var timeAverage = blockPercentageSeenByGatewayFirst*float64(blockReceivedByGrpcGatewayFirstAvgDelta) - (1-blockPercentageSeenByGatewayFirst)*float64(txReceivedByGatewayFirstAvgDelta)
+	results := fmt.Sprintf(
+		"\nAnalysis of Blocks received on both feeds:\n"+
+			"Number of blocks: %d\n"+
+			"Number of blocks received from gRPC connection first: %d\n"+
+			"Number of blocks received from websocket connection first: %d\n"+
+			"Percentage of blocks seen first from gRPC connection: %.2f%%\n"+
+			"Average time difference for blocks received first from gRPC connection (us): %d\n"+
+			"Average time difference for blocks received first from websocket connection (us): %d\n"+
+			"Final calculation (us): %.2f\n"+
+			"\nTotal Blocks summary:\n"+
+			"Total blocks from gRPC connection: %d\n"+
+			"Total blocks from ws connection: %d\n",
+		newBlockSeenByBothFeeds,
+		blockSeenByBothFeedsGrpcGatewayFirst,
+		blockSeenByBothFeedsGatewayFirst,
+		blockPercentageSeenByGatewayFirst*100,
+		blockReceivedByGrpcGatewayFirstAvgDelta,
+		txReceivedByGatewayFirstAvgDelta,
+		timeAverage,
+		totalBlockFromGrpcGateway,
+		totalBlockFromGateway)
+
+	verboseResults := fmt.Sprintf(
+		"Number of new blocks received first from gRPC connection: %d\n"+
+			"Number of new blocks received first from websocket connection: %d\n"+
+			"Total number of blocks seen: %d\n",
+		newBlockFromGrpcGatewayFeedFirst,
+		newBlockFromGatewayFeedFirst,
+		newBlockFromGrpcGatewayFeedFirst+newBlockFromGatewayFeedFirst,
+	)
+
+	if verbose {
+		results += verboseResults
+	}
+
+	return results
+}
+
+func (s *BlockGrpcWSCompareService) readFeedFromBX(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	out chan<- *message,
+	uri string,
+	authHeader string,
+	excludeBlockContents bool,
+) {
+	defer wg.Done()
+
+	log.Infof("Initiating connection to: %s", uri)
+	conn, err := ws.NewConnection(uri, authHeader)
+	if err != nil {
+		log.Errorf("cannot establish connection to %s: %v", uri, err)
+		return
+	}
+	log.Infof("Connection to %s established", uri)
+
+	defer func() {
+		if err := conn.Close(); err != nil {
+			log.Errorf("cannot close socket connection to %s: %v", uri, err)
+		}
+	}()
+
+	sub, err := conn.SubscribeBkFeedBX(1, s.feedName, excludeBlockContents)
+	if err != nil {
+		log.Errorf("cannot subscribe to feed %q: %v", s.feedName, err)
+		return
+	}
+
+	defer func() {
+		if err := sub.Unsubscribe(); err != nil {
+			log.Errorf("cannot unsubscribe from feed %q: %v", s.feedName, err)
+		}
+	}()
+
+	for {
+		var (
+			data, err    = sub.NextMessage()
+			timeReceived = time.Now()
+			msg          = &message{
+				bytes:        data,
```

Comment:
as you did for gRPC. you can send only the hash instead of the whole data, it could save some time

Reply:
changed